Mission download

julianoes opened the pull request:
This adds mission download and makes sure we compare the mission items uploaded and downloaded again.

julianoes committed:
Rename class MissionUpload to Mission

This way we can also add mission download to close the loop.

julianoes committed:
mission: refactoring and add mission download

julianoes committed:
Set line length to 100

For my taste 120 is way too long, especially for a 3 way merge.

julianoes committed:
mission: mark arguments unused

julianoes committed:
mission: implement comparison

julianoes committed:
mission: follow clang-tidy tips

julianoes committed:
cmake: clone yaml-cpp from GitHub into build folder

bkueng commented on src/tests/base.h:
```diff
@@ -7,6 +7,8 @@
 #include <memory>
 #include <string>
 
+#define UNUSED(x) (void)(x)
```
I'd rather disable the warning

julianoes replied:
Why? This makes it explicit.

bkueng replied:
I don't see why it needs to be explicit, I have not come across a case where an argument was accidentally ignored (not saying it doesn't happen though). Ignoring it due to inheritance or for other reasons is however quite common.

julianoes replied:
:thinking: right.

bkueng replied:
So what's your take on this?

bkueng commented on .clang-format:
```diff
@@ -1,6 +1,6 @@
 Language: Cpp
 BasedOnStyle: Google
-ColumnLimit: 120
+ColumnLimit: 100
```
I'm fine with it but:
- I don't want to have more-or-less arbitrary style changes based on personal preferences.
- This should preferably be a separate PR.

julianoes replied:
Okay!

bkueng reviewed (commented):
Looks good, thanks! One thing: can you remove the need for the `UNUSED` macro by disabling the warning instead?

julianoes committed:
tests: added expect assertion

This allows to have several checks per test but continue with the
remaining procedures.

bkueng commented on src/tests/mission.h:
```diff
@@ -5,13 +5,16 @@
 #include <dronecode_sdk/plugins/mission/mission.h>
 
+namespace dcsdk = dronecode_sdk;
+using mission_items = std::vector<std::shared_ptr<dcsdk::MissionItem>>;
```
Can you move this into the class?

julianoes replied:
I removed it.

bkueng commented on src/tests/base.cpp:
```diff
@@ -19,6 +22,28 @@ void TestBase::storeConfig(ConfigNode& config)
+{
+    std::stringstream ss(path);
+    std::string token;
+    while (std::getline(ss, token, '/'))
```
Why not using `std::string::find_last_of`?

julianoes replied:
That seems easier!

julianoes replied:
Implemented, thx!

julianoes committed:
cmake: add DronecodeSDK using ExternalProject_Add

This means that we can pin the version and prevent errors where we run
it against an outdated library version.

julianoes committed:
run_ci.sh: fixes after SDK is pulled in by cmake

bkueng commented on src/tests/base.h:
```diff
@@ -3,9 +3,45 @@
+
+#define UNUSED(x) (void)(x)
+
+// SFINAE to check if something is streamable adapted from:
```
Can you move this in some other helper/util header?

julianoes replied:
Ok.

julianoes replied:
Moved.

bkueng commented on src/tests/base.h:
```diff
@@ -62,9 +98,37 @@ class TestBase
+
+        auto maybe_streamable = MaybeStreamable<T>();
+
+        if (maybe_streamable.isStreamable()) {
```
Nice

bkueng commented on src/tests/base.h:
```diff
@@ -62,9 +98,37 @@ class TestBase
+                      << std::endl;
+        }
+
+        _result = Result::Failed;
```
I'd rather throw an exception here. Continue running a failed test does not make much sense to me.

julianoes replied:
It depends on the check. That's why usually there are `expects` and `asserts`. expects should just take note, and asserts should stop.

bkueng replied:
Fine, then we also need an assert.

julianoes replied:
Ok.

julianoes replied:
Implemented.

bkueng reviewed (commented):
I like the `EXPECT_EQ`, it simplifies the code quite a bit.

julianoes committed:
base: replace getline with find_last_of

This is easier and clang-tidy doesn't complain.

julianoes committed:
config: let's do more waypoints by default

julianoes committed:
cmake: remove leftover commented out line

julianoes committed:
cmake: only need to include ExternalProject once

julianoes committed:
cmake: move external projects into separate file

julianoes committed:
mission: remove using to shorten type

julianoes committed:
tests: move sfinae to its own header file

julianoes committed:
tests: add assert next to expect

If expects fail, the tests move on. If an assert fails, we stop.

bkueng commented on src/tests/base.cpp:
```diff
@@ -19,6 +22,25 @@ void TestBase::storeConfig(ConfigNode& config)
 
+std::string TestBase::extractFilename(const std::string& path)
+{
+    int pos = path.find_last_of('/');
```
To be on the safe side, can you handle `string::npos` as return value?

julianoes replied:
Will do.

julianoes committed:
base: abort if filename is unknown

bkueng committed:
Merge pull request #7 from Auterion/clone-sdk-automatically

cmake: add DronecodeSDK using ExternalProject_Add

julianoes commented:
And now we can merge this one and then I'll move to other things.

bkueng commented on src/tests/base.cpp:
```diff
@@ -19,6 +22,29 @@ void TestBase::storeConfig(ConfigNode& config)
 
+std::string TestBase::extractFilename(const std::string& path)
+{
+    auto pos = path.find_last_of('/');
```
I guess we should handle windows as well: '\\\\'

julianoes replied:
Fixed in 6af65e4.

bkueng commented on src/tests/mission.cpp:
```diff
@@ -6,49 +6,91 @@
-    }
-    cout << "Mission uploaded." << endl;
+    const dcsdk::Mission::Result result = fut.get();
+    EXPECT_EQ(result, dcsdk::Mission::Result::SUCCESS);
```
If the upload fails, there's probably no point in downloading it. Meaning this should be an assertion.

julianoes replied:
It might have been uploaded anyway but we got a timeout. I think we can still try to download it, and most probably it'll be wrong.

bkueng commented on src/tests/mission.cpp:
```diff
@@ -6,49 +6,91 @@
+void Mission::compareMissions(const std::vector<std::shared_ptr<dcsdk::MissionItem>>& items_a,
+                              const std::vector<std::shared_ptr<dcsdk::MissionItem>>& items_b)
+{
+    EXPECT_EQ(items_a.size(), items_b.size());
```
This needs to be an assertion. Otherwise you can access invalid data.

julianoes replied:
True.

julianoes replied:
Fixed in d27d0f8.

bkueng commented on src/tests/param.cpp:
```diff
@@ -14,43 +14,27 @@ REGISTER_TEST(ParamChange);
-    if (set_result2 != ParamsRaw::Result::SUCCESS) {
-        return Result::Failed;
-    }
+    EXPECT_EQ(set_result2, ParamsRaw::Result::SUCCESS);
```
Similar here: if setting the parameter fails, it's very likely that a read will fail as well (at least the value comparison).

julianoes replied:
Which is fine I think.

bkueng replied:
It's a question of what we want. It can be quite confusing, since you will see failures that are a consequence of other failures.

julianoes replied:
Right, so maybe we should use asserts more often as you suggest but then have them per test (e.g. Mission) only.

bkueng replied:
Exactly, that's what I was thinking.

julianoes replied:
Addressed in 2f8054d05e1fb33086702d51f749a1dd84ae67c5.

bkueng reviewed (commented):
I think most of the checks should be assertions.
I'm also wondering if the asserted exception should be caught, and continue with the next test, instead of aborting everything.

julianoes committed:
base: make path separator Windows compatible

julianoes committed:
mission: don't access outside bounds

julianoes committed:
base: catch asserts in test loop

bkueng reviewed (commented):
Not sure if you haven't gotten around to do so yet or forgot: can you have another pass at the code and replace EXPECT_EQ with ASSERT_EQ where needed? It's good to merge afterwards.

julianoes commented:
> Not sure if you haven't gotten around to do so yet or forgot: can you have another pass at the code and replace EXPECT_EQ with ASSERT_EQ where needed? It's good to merge afterwards.

Done, I think.
